mirrord/cli/src/port_forward.rs

```diff
@@ -0,0 +1,689 @@
+                        tracing::trace!("None message received, connection ended with agent");
+                        break Ok(())
+                    },
+                },
+
+                // stream coming from the user app
+                message = self.listeners.next() => match message {
+                    Some(message) => self.handle_listener_stream(message)?,
+                    None => break Ok(()),
+                },
+
+                message = self.internal_msg_rx.recv() => match message {
+                    Some(message) => self.handle_msg_from_task(message).await?,
+                    None => {
+                        tracing::trace!("PortForwardMessage Sender disconnected while sending");
+                    },
+                },
+            }
+        }
+    }
+
+    async fn handle_msg_from_agent(
+        &mut self,
+        message: DaemonMessage,
+    ) -> Result<(), PortForwardError> {
+        match message {
+            DaemonMessage::TcpOutgoing(message) => match message {
+                DaemonTcpOutgoing::Connect(res) => match res {
+                    Ok(res) => {
+                        let SocketAddress::Ip(remote_socket) = res.remote_address else {
+                            return Err(PortForwardError::ConnectionError(
+                                "Unexpectedly received Unix address for socket during setup".into(),
+                            ));
+                        };
+                        let Some(channel) = self.oneshots.pop_front() else {
+                            return Err(PortForwardError::ReadyTaskNotFound(
+                                remote_socket,
+                                res.connection_id,
+                            ));
+                        };
+                        self.sockets.insert(res.connection_id, remote_socket);
+                        let _ = channel.send(res.connection_id);
+                        tracing::trace!("Successful connection to remote address {remote_socket}, connection ID is {}", res.connection_id);
+                    }
+                    Err(error) => {
+                        return Err(PortForwardError::AgentError(format!(
+                            "Problem receiving DaemonTcpOutgoing::Connect {error}"
+                        )))
+                    }
+                },
+                DaemonTcpOutgoing::Read(res) => match res {
+                    Ok(res) => {
+                        let Some(&remote_socket) = self.sockets.get(&res.connection_id) else {
+                            return Err(PortForwardError::SocketFromIdNotFound(res.connection_id));
+                        };
+                        let Some(sender) = self.tasks.get(&remote_socket) else {
+                            return Err(PortForwardError::SenderNotFound(
+                                remote_socket,
+                                res.connection_id,
+                            ));
+                        };
```

Comment:
This can be an `unreachable!()`, the struct is encapsulated and you can ensure it never happens